### PR TITLE
Right-to-left tables

### PR DESCRIPTION
**Pull request by liZe**
Includes #1070 and #1071.

Related to #106.

### CODE REVIEW

**Comment by liZe** [comment]
@mohdnajdy Thanks a lot for your pull requests, I’ll add comments in the review.

File: weasyprint/layout/tables.py [hunk 1]
```diff
@@ -38,6 +37,10 @@ def table_layout(context, table, max_position_y, skip_stack, containing_block,
         position_x += width
     rows_width = position_x - rows_x
 
+    # This should fix the rtl direction for table
+    if table.style['direction'] == 'rtl':
+        column_positions.reverse()
+
```
Comment (liZe) [inline weasyprint/layout/tables.py]: This is actually wrong, reverting is not enough :/.

Let’s take a simple example: with a table with two columns and no borders, the first column’s width is 10px and the second one’s is 20px. If the table’s x position is 0, the ltr positions are (0, 10), and the rtl positions should be (20, 0).

```
LTR
0          10
+----------+--------------------+
| 1        | 2                  |
+----------+--------------------+

RTL
0                    20
+--------------------+----------+
| 2                  | 1        |
+--------------------+----------+
```

File: weasyprint/tests/test_draw/test_table.py [hunk 2]
```diff
@@ -531,3 +531,108 @@ def test_tables_11():
...
+        _gggggggggggggggggggggggggg_
+        ____________________________
+    '''), tables_source % {'extra_css': '''
+      x-table { border: solid #00f; border-width: 8px 2px;
+                table-layout: fixed; border-collapse: collapse;
+                direction: rtl }
+      x-td { border-color: #ff7f7f }
+      @page { size: 28px 26px; margin: 1px;
+              border: 1px solid rgba(0, 255, 0, 0.5); }
+    '''})
```
Comment (liZe) [inline weasyprint/tests/test_draw/test_table.py]: These tests are good! We just have to set rtl to the body, not the table, to get right-aligned tables.

File: weasyprint/tests/test_layout/test_table.py [hunk 3]
```diff
@@ -1536,6 +1536,36 @@ def test_layout_table_auto_50():
...
+    ''' % size)
+    html, = page.children
+    body, = html.children
+    wrapper, = body.children
+    table, = wrapper.children
+    row_group, = table.children
+
+    assert table.position_x == 0
+    assert int(table.width) == width
+
```
Comment (liZe) [inline weasyprint/tests/test_layout/test_table.py]: We have to test the positions of the columns as well.

**Comment by liZe** [comment]
I think that this PR is ready to merge.

We now have a decent rtl tables support, but also:
- `start` and `end` values allowed for `text-align`,
- better support of rtl text with floating blocks (not tested).

@mohdnajdy Feel free to test this branch with real-life documents and to give feedback if you have some time!

File: weasyprint/tests/test_layout/test_table.py (hunk 3 above)
Comment (mohdnajdy) [reply weasyprint/tests/test_layout/test_table.py]: I will test them and let you know.

**Comment by mohdnajdy** [comment]
Thanks @liZe i will test and let you know.

**Comment by liZe** [comment]
@mohdnajdy Did you find the time to test?

**Comment by mohdnajdy** [comment]
I will have time tomorrow to do so.

**Comment by mohdnajdy** [comment]
@liZe how can i get these changes in my fork ?

**Comment by liZe** [comment]
> @liZe how can i get these changes in my fork ?

The easiest way is to clone the repository in a separate folder and checkout the `rtl_tables` branch:

```
git clone https://github.com/Kozea/WeasyPrint.git
git checkout rtl_tables
```

**Comment by mohdnajdy** [comment]
---

@liZe I tested the rtl support for table and it's working nicely. I also tested the start, end alignment new values and they work fine.

---
but i have identified an issue i tried to write a test case for it but i am not confident in what i wrote since i still don't get what is B, _, r, p and these things are.

This is the test case.
```
@assert_no_logs
@requires('cairo', (1, 12, 0))
def test_tables_14():
    assert_pixels('table_row_backgrounds', 28, 28, to_pix('''
        ____________________________
        _BBBBBBBBBBBBBBBBBBBBBBBBBB_
        _B________________________B_
        _B________________________B_
        _B__bbbbbb_bbbbbb_bbbbbb__B_
        _B__bbbbbb_bbbbbb_bbbbbb__B_
        _B__bbbbbb_bbbbbb_bbbbbb__B_
        _B__bbbbbb_bbbbbb_bbbbbb__B_
        _B__bbbbbb_bbbbbb_bbbbbb__B_
        _B__bbbbbb_bbbbbb_bbbbbb__B_
        _B_________bbbbbb_________B_
        _B__bbbbbb_bbbbbbbpppppp__B_
        _B__bbbbbb_bbbbbbbpppppp__B_
        _B__bbbbbb_bbbbbbbpppppp__B_
        _B__bbbbbb_bbbbbbbpppppp__B_
        _B__bbbbbb_bbbbbbbpppppp__B_
        _B__bbbbbb_bbbbbbbpppppp__B_
        _B________________________B_
        _B_________rrrrrr_rrrrrr__B_
        _B_________rrrrrr_rrrrrr__B_
        _B_________rrrrrr_rrrrrr__B_
        _B_________rrrrrr_rrrrrr__B_
        _B_________rrrrrr_rrrrrr__B_
        _B_________rrrrrr_rrrrrr__B_
        _B________________________B_
        _B________________________B_
        _BBBBBBBBBBBBBBBBBBBBBBBBBB_
        ____________________________
    '''), tables_source % {'extra_css': '''
      body { direction: rtl }
      x-table { border-color: #00f; table-layout: fixed }
      x-tbody { background: rgba(0, 0, 255, 1) }
      x-tr { background: rgba(255, 255, 255, 255) }
    '''})
```
basiclly when i tried to render background colors for rtl it fails. this is on the rtl_tables branch.

The html i used is this
```
<!DOCTYPE html>
<html>
<head>
<style>
body {
  direction: rtl;
}  
table {
  border-collapse: collapse;
}

td, th {
  border: 1px solid #dddddd;
  padding: 8px;
}

tr:nth-child(even) {
  background-color: #dddddd;
}
</style>
</head>
<body>

<h2>HTML Table</h2>

<table>
  <tr>
    <th>الاسم الأول</th>
    <th>الاسم الثاني</th>
    <th>العائلة</th>
  </tr>
  <tr>
    <td>محمد</td>
    <td>يوسف</td>
    <td>النجدي</td>
  </tr>
  <tr>
    <td>محمد</td>
    <td>يوسف</td>
    <td>النجدي</td>
  </tr>
  <tr>
    <td>محمد</td>
    <td>يوسف</td>
    <td>النجدي</td>
  </tr>
</table>

</body>
</html>
```
I narrow down of where is this happening and i think it's from this block of code.
```
        # Background color
        if bg.color.alpha > 0:
            with stacked(context):
                painting_area = bg.layers[-1].painting_area
                if painting_area:
                    if bleed:
                        # Painting area is the PDF BleedBox
                        x, y, width, height = painting_area
                        painting_area = (
                            x - bleed['left'], y - bleed['top'],
                            width + bleed['left'] + bleed['right'],
                            height + bleed['top'] + bleed['bottom'])
                    context.rectangle(*painting_area)
                    context.clip()
                context.set_source_rgba(*bg.color)
                context.paint()
```

**Comment by mohdnajdy** [comment]
@liZe i feel like after such changes all table tests should have a sub-test for RTL to be more confident that such changes will work as expected like before or not.

**Comment by mohdnajdy** [comment]
@liZe 
for this method i think if a table element do not have background it should be skipped.
```
def draw_table_backgrounds(context, page, table, enable_hinting):
            """ other code """
            draw_background(context, row.background, enable_hinting) # from this
            if row.background: # To this
                draw_background(context, row.background, enable_hinting) # To this
            """ other code """
```
what do you think ?

**Comment by liZe** [comment]
> I tested the rtl support for table and it's working nicely. I also tested the start, end alignment new values and they work fine.

Good to know, thanks!

> what do you think ?

`draw_background` returns early when there’s no background to render. It’s better than to duplicate the check outside the function, even if it costs a function call each time.

> but i have identified an issue i tried to write a test case for it but i am not confident in what i wrote since i still don't get what is B, _, r, p and these things are.

These variables are [colors](https://github.com/Kozea/WeasyPrint/blob/master/weasyprint/tests/test_draw/__init__.py), each letter represents a pixel.

> basiclly when i tried to render background colors for rtl it fails.

You’re right, I’ll fix this and add a test for this. Thank you for reporting the issue.

> i feel like after such changes all table tests should have a sub-test for RTL to be more confident that such changes will work as expected like before or not.

Well, that’s a lot of work… I’d really like to have these tests, but I’m afraid I don’t have the time to duplicate all of them. If anyone is interested, I’ll be happy to merge them :smile:.

**Comment by mohdnajdy** [comment]
> Well, that’s a lot of work… I’d really like to have these tests, but I’m afraid I don’t have the time to duplicate all of them. If anyone is interested, I’ll be happy to merge them 😄.

@liZe i am really in. I will start with table tests if that is fine.

File: weasyprint/tests/test_draw/test_table.py [hunk 4]
```diff
@@ -281,6 +281,45 @@ def test_tables_5():
...
+        _B_________rrrrrr_rrrrrr__B_
+        _B_________rrrrrr_rrrrrr__B_
+        _B_________rrrrrr_rrrrrr__B_
+        _B_________rrrrrr_rrrrrr__B_
+        _B________________________B_
+        _B________________________B_
+        _BBBBBBBBBBBBBBBBBBBBBBBBBB_
+        ____________________________
+    '''), tables_source % {'extra_css': '''
+      x-table { border-color: #00f; table-layout: fixed; direction: rtl }
```
Comment (mohdnajdy) [inline weasyprint/tests/test_draw/test_table.py:317]: @liZe will we keep the direction at the table level or it should be on the body ?

Comment (liZe) [reply weasyprint/tests/test_draw/test_table.py:317]: > @liZe will we keep the direction at the table level or it should be on the body ?

Putting `direction` on the table avoids a right-aligned table but keeps rtl columns and cells, that’s what we want to test.

**Comment by mohdnajdy** [comment]
@liZe i wrote tests which covers rtl for tables. basically duplicated them. 2 of them failed. 
I want to push these changes to this branch but i can't

**Comment by mohdnajdy** [comment]
#1097 added more test cases here

**Comment by liZe** [comment]
> @liZe i wrote tests which covers rtl for tables. basically duplicated them. 2 of them failed.
> I want to push these changes to this branch but i can't

:heart:

Thanks a lot, I’ll merge them both and fix failing tests.

**Pull request closed without merging**